### PR TITLE
ORSP: Convert Data Use Restriction Edit

### PR DESCRIPTION
**Pull request by vvicario**
## Addresses
https://broadinstitute.atlassian.net/browse/BTRX-749

## Changes
Added new component for data use restriction
## Testing
Go to Consent Group -> Documents -> Upload a document with type Data Use Limitation -> Create Restriction. Restriction should be created.
After creation try to edit.
---

- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Get a thumb from Belatrix
- [ ] **Submitter**: Get a thumb from @rushtong
- [ ] **Submitter**: Merge to develop using github's "Squash and Merge" feature
- [ ] **Submitter**: Delete branch after merge

### CODE REVIEW

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy [hunk 1]
```diff
@@ -89,6 +94,19 @@ class DataUseController extends AuthenticatedController {
         }
     }
 
+    def createSdul() {
+        try {
+            DataUseRestriction restriction = DataUseRestriction.findByConsentGroupKey(request.JSON.consentGroupKey)
+            restriction = DataUseRestrictionParser.fromParams(restriction, request.JSON)
+            dataUseLetterService.createSdul(restriction, getUser()?.displayName)
+            response.status = 200
```
Comment (dags) [inline grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy:88]: Should be 201 ?

File: src/main/webapp/components/Documents.js [hunk 2]
```diff
@@ -11,9 +11,18 @@ import './Documents.css';
 import { UrlConstants } from "../util/UrlConstants";
 import { spinnerService } from "../util/spinner-service";
 import { Spinner } from "./Spinner";
+import { Link } from 'react-router-dom';
+import '../components/Btn.css';
```
Comment (dags) [inline src/main/webapp/components/Documents.js]: Do we still need this css ?

File: src/main/webapp/components/MultiSelect.js [hunk 3]
```diff
@@ -87,6 +87,7 @@ export const MultiSelect = hh(class MultiSelect extends Component {
       }, [
           div({ className: "inputFieldSelectWrapper" }, [
             h(AsyncSelect, {
+              key:this.props.id,
```
Comment (dags) [inline src/main/webapp/components/MultiSelect.js]: I think this is not required.

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy [hunk 4]
```diff
@@ -89,6 +94,19 @@ class DataUseController extends AuthenticatedController {
         }
     }
 
+    def createSdul() {
+        try {
+            DataUseRestriction restriction = DataUseRestriction.findByConsentGroupKey(request.JSON.consentGroupKey)
+            restriction = DataUseRestrictionParser.fromParams(restriction, request.JSON)
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: These two construction lines are a little confusing. `request.JSON.consentGroupKey` can fail if JSON doesn't have the right property. If it isn't found, this should probably return a 404. Secondly, it's strange to see a find immediately followed by a reassignment. I can follow the code to see that this is getting the original restriction, then updating it from the passed in parameters. What this should look like is a dedicated service method that `finds and updates a restriction from parameters` which then returns the mutated/unsaved restriction.

File: grails-app/utils/org/broadinstitute/orsp/utils/DataUseRestrictionParser.groovy [hunk 5]
```diff
@@ -10,9 +10,9 @@ class DataUseRestrictionParser {
         dataUseRestriction = dataUseRestriction == null ? new DataUseRestriction() : dataUseRestriction
         dataUseRestriction.consentGroupKey = params.consentGroupKey
         dataUseRestriction.consentPIName = params.consentPIName
-        dataUseRestriction.generalUse = IssueUtils.getBooleanForParam(params.generalUse)
-        dataUseRestriction.hmbResearch = IssueUtils.getBooleanForParam(params.hmbResearch)
-        dataUseRestriction.manualReview = IssueUtils.getBooleanForParam(params.manualReview)
+        dataUseRestriction.generalUse = params.generalUse instanceof Boolean ? params.generalUse : IssueUtils.getBooleanFromString(params.generalUse)
+        dataUseRestriction.hmbResearch = params.hmbResearch instanceof Boolean ? params.hmbResearch : IssueUtils.getBooleanFromString(params.hmbResearch)
+        dataUseRestriction.manualReview = params.manualReview instanceof Boolean ? params.manualReview : IssueUtils.getBooleanFromString(params.manualReview)
```
Comment (rushtong) [inline grails-app/utils/org/broadinstitute/orsp/utils/DataUseRestrictionParser.groovy]: Pretty sure this looks like an over-complication of something that should be very simple. Instead of making all of these changes here, why not just handle the `instanceof` in `IssueUtils.getBooleanForParam`?

File: grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy [hunk 6]
```diff
@@ -103,7 +103,7 @@ final class IssueUtils {
 
     }
 
-    static Boolean getBooleanForParam(String param) {
+    static Boolean getBooleanFromString(String param) {
```
Comment (rushtong) [inline grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy]: Suggestion: revert the name change and add the boolean check first, i.e. add a line above the two ifs:
```
if (param instanceof Boolean) return param
```
That one line saves you 15 unnecessary edits.

File: src/main/webapp/components/Documents.js (hunk 2 above)
Comment (rushtong) [reply src/main/webapp/components/Documents.js]: We should be asking that question for all CSS. If we can inline style something, do it. Shared CSS is a huge headache. We should try to limit any css classes to the standard bootstrap stuff we use.

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 7]
```diff
@@ -0,0 +1,864 @@
...
+import { h1, h, div, hh, h2, strong, ul, i, p, li, label, abbr, input, span, hr, button } from 'react-hyperscript-helpers';
+import { InputFieldText } from "../components/InputFieldText";
+import { InputYesNo } from "../components/InputYesNo";
+import { Search } from "../util/ajax";
+import { InputFieldTextArea } from "../components/InputFieldTextArea";
+import { MultiSelect } from "../components/MultiSelect";
+import { AlertMessage } from '../components/AlertMessage';
+import { DataUse, ConsentGroup } from "../util/ajax";
+import { isEmpty } from '../util/Utils';
+import { UrlConstants } from '../util/UrlConstants';
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js]: unused

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 8]
```diff
@@ -0,0 +1,864 @@
...
+    super(props);
+    let params = new URLSearchParams(this.props.location.search);
+    this.state = {
+      consent: {
+        projectKey: '',
+        summary: ''
+      },
+      showError: false,
+      restriction: this.initRestriction(),
+      create: params.get('create') !== undefined ? true : false,
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js]: Can just be `create: params.get('create') !== undefined,` 
I also just learned about [lodash isNil](https://lodash.com/docs/4.17.15#isNil)

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 9]
```diff
@@ -0,0 +1,864 @@
...
+      publicationResults: restriction !== undefined ? restriction.publicationResults : '',
+      genomicResults: restriction !== undefined ? restriction.genomicResults : '',
+      geographicalRestrictions: restriction !== undefined ? restriction.geographicalRestrictions : '',
+      other: restriction !== undefined ? restriction.other : '',
+      manualReview: restriction !== undefined ? restriction.manualReview : false,
+      comments: restriction !== undefined ? restriction.comments : '',
+      populationRestrictions: restriction !== undefined ? this.getAutocompleteData(restriction.populationRestrictions) : [],
+      genomicSummaryResults: restriction !== undefined && !isEmpty(restriction.genomicSummaryResults) ? restriction.genomicSummaryResults : '',
+      genomicPhenotypicData: restriction !== undefined ? restriction.genomicPhenotypicData : '',
+      consentPIName: restriction !== undefined && !isEmpty(restriction.consentPIName) ? restriction.consentPIName : '',
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js]: This wall of `!== undefined` ternaries can be made more legible with some [lodash get](https://lodash.com/docs/4.17.15#get). For example this should work (untested), 
```
consentPIName: _.get(restriction, 'consentPIName', '')
```

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 10]
```diff
@@ -0,0 +1,864 @@
...
+    };
+    return resp;
+  }
+
+  init() {
+    spinnerService.show(RESTRICTION_SPINNER);
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js:94]: Note for @LeoForconesi - this might impact your spinner PR.

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 11]
```diff
@@ -0,0 +1,864 @@
...
+        prev.restriction.diseaseRestrictions = [];
+        prev.restriction.generalUse = true;
+        return prev;
+      }, () => this.validateForm());
+    } else {
+      this.setFieldValue(field, value);
+    }
+  }
+
+  handleGenomicResultChange = (e, field, value) => {
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js]: Looks unused

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 12]
```diff
@@ -0,0 +1,864 @@
...
+        prev.restriction.populationOriginsAncestry = true;
+        prev.restriction.hmbResearch = true;
+        return prev;
+      }, () => this.validateForm());
+    } else {
+      this.setFieldValue(field, value);
+    }
+  }
+
+  checkIfExist(data, key) {
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js]: Looks unused

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js [hunk 13]
```diff
@@ -0,0 +1,864 @@
...
+            })
+          ]),
+          hr({ isRendered: this.state.restriction.genomicResults }, []),
+          div({ isRendered: this.state.restriction.genomicResults, style: styles.inputGroup }, [
+            p({ className: "inputFieldLabel" }, ["If ", i({}, ["Yes"]), ", please explain."]),
+            InputFieldTextArea({
+              disabled: false,
+              value: this.state.restriction.genomicSummaryResults,
+              name: 'genomicSummaryResults',
+              onChange: this.changeTextHandler
```
Comment (rushtong) [inline src/main/webapp/dataUse/DataUseRestrictionEdit.js:791]: Maybe here and for the input above you want `handleGenomicResultChange` ?

File: src/main/webapp/main/Routes.js [hunk 14]
```diff
@@ -28,6 +30,9 @@ const Routes = ( props ) => (
     <Route path= {"/dataUseLetter/show"} render = {(routeProps) =>  <DataUseLetterIndex {...routeProps} {...props}/> }/>
     <Route path= {"/submissions/add-new"} render={(routeProps) => <SubmissionForm {...routeProps} {...props} />} />
     <Route path= {"/dataUse/show"} render = {(routeProps) =>  <DataUseRestrictionDetails {...routeProps} {...props}/> }/>
+
+    <Route path= {"/dataUse/restriction"} render = {(routeProps) =>  <DataUseRestrictionEdit {...routeProps} {...props}/> }/>
+
```
Comment (rushtong) [inline src/main/webapp/main/Routes.js]: nit: remove extra lines surrounding this route

File: src/main/webapp/util/UrlConstants.js [hunk 15]
```diff
@@ -63,21 +64,24 @@ export const UrlConstants = {
...
   submissionSaveUrl: context + '/api/submissions/save-new',
   submissionDisplayUrl: context + '/api/submissions/display',
   submissionRemoveFileUrl: context + '/api/submission/remove-file',
-
   sampleConsentLinkUrl: context + '/api/sample-consent-link',
   sampleBreakLinkUrl: context + '/api/break-link',
   sampleApproveLinkUrl: context + '/api/approve-link',
   historyUrl: context + '/api/history',
+  submissionsUrl: context + '/api/submissions',
+  submissionDisplayUrl: context + '/api/submissions/display',
```
Comment (rushtong) [inline src/main/webapp/util/UrlConstants.js]: Lines 80 and 81 are duplicates

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy [hunk 16]
```diff
@@ -22,14 +23,18 @@ class DataUseController extends AuthenticatedController {
...
+
+    def restriction() {
+      redirectToMainContainer()
     }
 
-    @Override
     def show() {
-        render(view: "/mainContainer/index", model:[restrictionId: params.id])
+      redirectToMainContainer()
     }
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: Why do we need three methods to do the same exact thing? We should have a single method, `index` (or whatever, I don't really care) that does what `redirectToMainContainer` does. I suggest we revert the change in `AuthenticatedController`, delete these three, and then change whatever GSP or JS file tries to use these methods. We need to trim outdated code, not support it.

**Review by rushtong — CHANGES_REQUESTED** [review]
See comments inline. Caveat, I did not test this page locally, just code reviewed. There is a lot of complicated logic in the original that I hope is preserved here, but this will need a review from the ORSP team for full functionality testing.

File: grails-app/utils/org/broadinstitute/orsp/utils/DataUseRestrictionParser.groovy (hunk 5 above)
Comment (dags) [reply grails-app/utils/org/broadinstitute/orsp/utils/DataUseRestrictionParser.groovy]: done

File: grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy (hunk 6 above)
Comment (dags) [reply grails-app/utils/org/broadinstitute/orsp/utils/IssueUtils.groovy]: done

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js (hunk 7 above)
Comment (dags) [reply src/main/webapp/dataUse/DataUseRestrictionEdit.js]: done

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 16 above)
Comment (vvicario) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: I tried to do that but there is an issue. If we have different url in URLMappings with the same controller and action, it executes the first match, and we need different url to be handled by React router. If it executes the first url that it's found, then router  doesn't return the expected component. 
we will work on that in the following story:
https://broadinstitute.atlassian.net/browse/BTRX-637

Comment (rushtong) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: Why would you have:
> If we have different url in URLMappings with the same controller and action

? I'm not suggesting that you do that. In fact, I would argue that that is an error.

File: src/main/webapp/dataUse/DataUseRestrictionEdit.js (hunk 8 above)
Comment (vvicario) [reply src/main/webapp/dataUse/DataUseRestrictionEdit.js]: Nice

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 16 above)
Comment (LeoForconesi) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: Let me jump in here. 
Since Groovy is catching every route change before react (eg. when refreshing the browser's page), we needed a way to redirect these calls to our MainContainer view (where React router lives) using `redirectToMainContainer()`, which, in this case is defined in DataUseController. 
So in order to do this redirection, and match the url with the one present in react router we created methods that match with them when groovy gets these urls.
We used this only for pages resfresh, and return to previous page navigation.

Comment (rushtong) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: > Since Groovy is catching every route change before react (eg. when refreshing the browser's page)

Agreed. What I'm suggesting is that all of the various calls to the these methods be consolidated into a single API call. Those page calls can come from JS or GSP, it doesn't matter. It makes no sense to support three different code paths to the exact same behavior.

**Comment by vvicario** [comment]
@rushtong can you re-review please?

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 1 above)
Comment (vvicario) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy:88]: I renamed this method to save instead of create

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy [hunk 17]
```diff
@@ -89,6 +93,18 @@ class DataUseController extends AuthenticatedController {
         }
     }
 
+    def saveSdul() {
+        try {
+            DataUseRestriction restriction = dataUseLetterService.updateRestrictionFromParams(request.JSON)
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: This method now throws an Illegal Argument exception, which should probably return a 400 bad request. 

From the previous round of feedback, it should also check for existence, which would return a 404 if not found.

File: grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy [hunk 18]
```diff
@@ -1,6 +1,8 @@
 package org.broadinstitute.orsp
 
 import grails.gorm.transactions.Transactional
+import javassist.NotFoundException
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy]: unused

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy [hunk 19]
```diff
@@ -89,6 +93,18 @@ class DataUseController extends AuthenticatedController {
...
+            dataUseLetterService.saveRestriction(restriction, getUser()?.displayName)
+            response.status = 200
+            render(response.status)
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: In general REST practice, when an API saves an object, it should return the object just saved. That way callers can be sure the save really happened.

File: grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy [hunk 20]
```diff
@@ -56,5 +58,13 @@ class DataUseLetterService {
         restriction
     }
 
+    DataUseRestriction updateRestrictionFromParams(Object params) {
+        if (params.consentGroupKey) {
+            DataUseRestriction existentRestriction = DataUseRestriction.findByConsentGroupKey(params.consentGroupKey)
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy:59]: Should probably throw the not found exception here if not found.

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 21]
```diff
@@ -400,17 +400,21 @@ class QueryService implements Status {
...
+                    ' select * from storage_document ' +
+                            ' where consent_collection_link_id in :consentCollectionIds' +
+                            ' and deleted = 0'
+            final SQLQuery sqlQuery = session.createSQLQuery(query)
+            results = sqlQuery.with {
+                addEntity(StorageDocument)
+                setParameterList('consentCollectionIds', consentCollectionId)
+                list()
+            }
         }
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: You don't need the redundant variable `results`. Just let the sqlQuery be the last line and then add an else with `Collections.emptyList()` so either the query results or the empty list is returned.

File: grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy (hunk 18 above)
Comment (dags) [inline grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy]: unused. Please remove it.

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 16 above)
Comment (rushtong) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: Doesn't look like this feedback was addressed.

**Review by rushtong — CHANGES_REQUESTED** [review]
See comments inline.

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 17 above)
Comment (vvicario) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: If consent group doesn't have a restriction it will be created in DataUseRestrictionParser. This is why I am not returning 404. If exists it uses the existent one. Does this clarify the logic?
dataUseRestriction = dataUseRestriction == null ? new DataUseRestriction() : dataUseRestriction

File: grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy (hunk 20 above)
Comment (vvicario) [reply grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy:59]: If it is not found, it will be created in DataUseRestrictionParser

Comment (rushtong) [reply grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy:59]: Gotcha, yes, thank you!

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 17 above)
Comment (rushtong) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy]: For the 404 question, yes, this makes sense that you don't need to handle not-found. The 400 handling is still not correct though, I'll comment again in the updated code.

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy [hunk 22]
```diff
@@ -89,6 +81,18 @@ class DataUseController extends AuthenticatedController {
         }
     }
 
+    def saveSdul() {
+        try {
+            DataUseRestriction restriction = dataUseLetterService.getRestrictionFromParams(request.JSON)
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy:86]: Follow-up from previous feedback. This method can throw an illegal argument exception which is converted into a 500 below. We should catch that case and return a BadRequest (400) instead.

File: grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy [hunk 23]
```diff
@@ -42,9 +43,9 @@ class DataUseLetterService {
         }
     }
 
-    DataUseRestriction createSdul(DataUseRestriction restriction, String displayName) {
+    DataUseRestriction saveRestriction(DataUseRestriction restriction, String displayName) {
         Issue consent = queryService.findByKey(restriction.consentGroupKey)
-        def updatedOrCreated = "Created"
+        def updatedOrCreated = "Updated"
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/DataUseLetterService.groovy]: This is previous code that I missed, sorry. Avoid `def` when possible. Also, this could be a lot simpler:
```
String updatedOrCreated = (restriction.id == null) ? "Created" : "Updated"
```

File: grails-app/utils/org/broadinstitute/orsp/utils/DataUseRestrictionParser.groovy [hunk 24]
```diff
@@ -54,7 +54,7 @@ class DataUseRestrictionParser {
         dataUseRestriction.recontactMust = params.recontactMust
         dataUseRestriction.genomicPhenotypicData = params.genomicPhenotypicData
         dataUseRestriction.cloudStorage = params.cloudStorage
-        dataUseRestriction.irb = IssueUtils.getBooleanForParam(params.irb)
+        dataUseRestriction.irb =  IssueUtils.getBooleanForParam(params.irb)
```
Comment (rushtong) [inline grails-app/utils/org/broadinstitute/orsp/utils/DataUseRestrictionParser.groovy]: nit: extraneous spacing, this change can be reverted

File: src/main/webapp/PageNotFound/PageNotFound.js [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+     return (
+       div({className:"row", style: {"margin":"60px 0"}}, [
+        div({className:"col-xs-12"}, [
+          h1({}, ["Sorry, this page could not be found."])
+         ])
+       ])
+     )
+  }
+})
+export default NotFound;
```
Comment (rushtong) [inline src/main/webapp/PageNotFound/PageNotFound.js:21]: This feedback doesn't necessarily need to be applied, but this component can be a [pure functional component](https://logrocket.com/blog/pure-functional-components/) instead. There's no props or state that are used in the render (although those can be used in functional components too...). Something like this should work and be a lot fewer lines of code:
```
export const NotFound = () => {
  return (
    div({className:"row", style: {"margin":"60px 0"}}, [
     div({className:"col-xs-12"}, [
       h1({}, ["Sorry, this page could not be found."])
      ])
    ])
  );
};

```

File: src/main/webapp/main/Routes.js [hunk 26]
```diff
@@ -28,6 +30,8 @@ const Routes = ( props ) => (
     <Route path= {"/dataUseLetter/show"} render = {(routeProps) =>  <DataUseLetterIndex {...routeProps} {...props}/> }/>
     <Route path= {"/submissions/add-new"} render={(routeProps) => <SubmissionForm {...routeProps} {...props} />} />
     <Route path= {"/dataUse/show"} render = {(routeProps) =>  <DataUseRestrictionDetails {...routeProps} {...props}/> }/>
+    <Route path= {"/dataUse/restriction"} render = {(routeProps) =>  <DataUseRestrictionEdit {...routeProps} {...props}/> }/>
+    <Route path= {"/*"} render = {(routeProps) =>  <PageNotFound {...routeProps} {...props}/> }/>
```
Comment (rushtong) [inline src/main/webapp/main/Routes.js:34]: The not found component is a recent addition to this PR. Has `path= {"/*"}` been tested? I'm thinking it won't have any effect on the non-react pages, but we should be sure.

**Review by rushtong — CHANGES_REQUESTED** [review]
One minor change requested which is to properly handle the 400 case. Also, the Not Found component is an unnecessary addition to this PR. It really should be in it's own ticket and not part of this one.

File: src/main/webapp/main/Routes.js (hunk 26 above)
Comment (dags) [reply src/main/webapp/main/Routes.js:34]: This is intended to work in combination with urlMappings change. Anything not resolved in urlMappings is redirected to be handled by Routes. And if not found there a 404 page is shown. It was tested late yesterday and worked as expected, with a still not resolved issue of getting an additional 500 error in notFound page. This change is related to handle yor comment about "redirectToMainContainer" .

File: grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy (hunk 22 above)
Comment (dags) [reply grails-app/controllers/org/broadinstitute/orsp/api/DataUseController.groovy:86]: additional catch added

File: src/main/webapp/main/Routes.js (hunk 26 above)
Comment (rushtong) [reply src/main/webapp/main/Routes.js:34]: The conversation around `redirectToMainContainer` was related to having three different methods performing the same function in a Groovy controller method. I don't see the connection to this. In either case, it's a relatively small concern.

**Review by rushtong — APPROVED** [review]

**Review by dags — APPROVED** [review]

File: src/main/webapp/main/Routes.js (hunk 26 above)
Comment (dags) [reply src/main/webapp/main/Routes.js:34]: Will try to explain it. We had 3 methods performing same function on Groovy controller because were handling 3 different urls (vies, pages). Since those pages are implemented in React now, but called from Grails by now, we need to go to the Grails controller and redirect to React Router which will handle the request at the end. While the 3 methods looked the same, redirectToMainContainer call was different on each case because the payload: the requested URL. That requested URL ended in Routes and was handled by Switch and Router. So, we removed actions from Grails controller and add a "generic" urlmapping in urlMappings to handle any not-matcher url and redirect it to React Router. Since any non existing url will end in react router as well, we added the /* route at end to catch all non exiting pages. That is the idea. Not sure if the best, at least the best we had. :-)

Comment (rushtong) [reply src/main/webapp/main/Routes.js:34]: I think this might be the disconnect ...

> Since those pages are implemented in React now, but called from Grails by now

If we have pages implemented in React, then there should be no calls from Grails to those pages. It sounds like there's a circular route handling process here that could use some de-convoluting. I wish I could spend some time simplifying this ... just not possible right now.

Comment (dags) [reply src/main/webapp/main/Routes.js:34]: I agree, there is something that needs improvement. We already have some stories in this Sprint for this, converting remaining pages/views to React.
